csharp/span_based_serialization.md

```diff
@@ -0,0 +1,63 @@
+in terms of the new .NET Span APIs in a way that requires no copying of binary payloads and basically allows deserialization directly
+from the buffer segments that were received from the network (and serializing directly into the buffer that is going to be sent out
+over the network).
+
+## Objective
+The goals for this effort:
+- keep the existing CodedInputStream and CodedOutputStream (no plans to remove them in the future), serialization/deserialization speed of these APIs should stay unaffected.
+- all the changes made need to be backward compatible
+- introduce a set of new public APIs for serialization/deserialization that utilize the Span type
+- allow future incremental performance optimizations and make them as easy as possible
+- ensure the are microbenchmark that allow comparing the two implementations
+- as support for the new serialization/deserialization APIs requires generating code that is not compatible with some older .NET frameworks, figure out a schema under which messages with "new serialization API support" and without it can coexist (without causing user friction).
+- the effort to maintain the Protobuf C# project should remain approximately the same (should not be negatively affected by introduction of new serialization/deserialization APIs).
+
+Non-goals
+- over-optimize the initial version of new serializer/deserializer (we should rather focus on enabling future improvements)
+- remove or deprecate the existing serialization/deserialization API
+
+## The Proposal
+
+Key elements:
+- Add CodedInputReader and CodedOutputWriter types (both are `ref structs` and thus can hold an instance of `Span`). These two types implement the low-level wire format semantics in terms of the `Span` type (as opposed to `byte[]` used by CodedInputStream and CodedOutputStream).
+- Introduce `IBufferMessage` interface (exact name TBD) which inherits from `IMessage` and marks protobuf messages that support the new serialization/deserialization API.
+- Introduce a new codegen option (exact name TBD) to enable/disable generating code required by CodedInputReader/CodedInputWriter (the MergeFrom and WriteTo methods). This option will default to disabled. gRPC frameworks will likely provide the ability to explicitly control it, e.g. `<Protobuf Include="..\Shared\benchmark_service.proto" Serialization="Buffer" />`, and/or detect the version of .NET being targeted and automatically enable the option.
+
+A prototype implementation is https://github.com/protocolbuffers/protobuf/pull/5888
+
+## Platform support
+
+- the new serialization API will be supported for projects that support at least `netstandard2.0` and allow use of C# 7.2. On newer .NET frameworks, there might be (and likely will be) further performance benefits.
+
+## Backwards compatibility
+
+## Code duplication
+
+TODO: testing (how to test both APIs without twice the effort).
+
+## Other considerations
```

Comment:
New consideration: can we do something about  parsing / serialization of `bytes` fields? The issue is that if the message contains large ByteString fields, the new parsing API is not going to help much because all the data will be copied into newly allocated `byte[]` objects. To some extent this is unavoidable, but we should at least give this some thought. AFAIK `bytes` fields are fairly common in some use cases.  e.g. for a hypothetical `FileDownloadService`, is there a way one could avoid the LOH allocations easily?

Reply:
The new design will save one allocation/copy. Optimizing ByteString is a separate allocation/copy problem.

Renting the underlying storage is one option. However that would require a way for devs to say they are releasing the message.

Create a new issue for this.

Reply:
Some solutions to eliminating large ByteString allocations: https://github.com/protocolbuffers/protobuf/issues/4206#issuecomment-567350035

I think this is an issue to look at separately.